From 1dc468db44ef6c4249476789c441b4d6a95323e1 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sun, 31 Jan 2021 18:15:33 +0100
Subject: [PATCH 091/179] ARM: dovetail: enable I-pipe compat for syscall
 routing

Allow legacy applications to issue syscalls following the I-pipe
syscall convention in EABI mode, which specifies that r7 should be
loaded with 0xf0042 in order to mark out-of-band syscalls.

When CONFIG_IPIPE_COMPAT is enabled, OOB_SYSCALL_BIT is ORed into r7
on the fly if its original value is 0xf0042, so that the syscall will
be routed to the companion core as expected.

This compat mode may be removed on short notice. Do NOT rely on it for
new applications.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 arch/arm/Kconfig               | 4 ++++
 arch/arm/kernel/entry-common.S | 9 +++++++++
 2 files changed, 13 insertions(+)

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index 24e6f70af..73eb1ee6f 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -235,6 +235,10 @@ config NEED_RET_TO_USER
 config ARCH_MTD_XIP
 	bool
 
+# Limited I-pipe compat (syscall routing only).
+config IPIPE_COMPAT
+	bool
+
 config ARM_PATCH_PHYS_VIRT
 	bool "Patch physical to virtual translations at runtime" if EMBEDDED
 	default y
diff --git a/arch/arm/kernel/entry-common.S b/arch/arm/kernel/entry-common.S
index 75d22a6a0..effd613a3 100644
--- a/arch/arm/kernel/entry-common.S
+++ b/arch/arm/kernel/entry-common.S
@@ -283,9 +283,18 @@ ENTRY(vector_swi)
 local_restart:
 #ifdef CONFIG_DOVETAIL
 	ldr	r10, [tsk, #TI_LOCAL_FLAGS]	@ tsk(r9) is callee-saved
+#ifdef CONFIG_IPIPE_COMPAT
+	ldr	r0, =#0xf0042			@ old syscall signature
+	cmp	scno, r0
+	bne	1f
+	or	scno, scno, #__OOB_SYSCALL_BIT	@ force in oob marker
+	b	fastcall_entry
+1:
+#endif	
 	ldr	r0, =#__OOB_SYSCALL_BIT
 	ands	r0, scno, r0
 	beq	slow_path
+fastcall_try:
 	tst	r10, #_TLF_OOB
 	beq	slow_path
 	mov	r0, sp				@ regs
-- 
2.38.1

